File: app/src/androidTest/java/ru/iteco/fmhandroid/ui/AppActivityTest.java
package ru.iteco.fmhandroid.ui;

import static ru.netology.data.DataHelper.checkMessage;

import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.filters.LargeTest;

import org.junit.Before;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import io.qameta.allure.android.runners.AllureAndroidJUnit4;
import io.qameta.allure.kotlin.Description;
import io.qameta.allure.kotlin.junit4.DisplayName;
import ru.iteco.fmhandroid.R;
import ru.iteco.fmhandroid.ui.AppActivity;
import ru.netology.data.DataHelper;
import ru.netology.data.ReadyScreen;
import ru.netology.pages.AuthPage;
import ru.netology.pages.MainScreenPage;

@LargeTest
@RunWith(AllureAndroidJUnit4.class)

public class AuthUnitTest {
    AuthPage authPage = new AuthPage();
    MainScreenPage mainScreenPage = new MainScreenPage();
    ReadyScreen readyScreen = new ReadyScreen();

    @Rule
    public ActivityScenarioRule<AppActivity> mActivityScenarioRule =
            new ActivityScenarioRule<>(AppActivity.class);

    @Before
    public void readyScreen() {
        readyScreen.readyAuthScreen();
    }

    @Test
    @DisplayName("2.Вход в личный кабинет с валидными данными")
    public void testAuthRight() {
        authPage.enterLogin(DataHelper.AuthInfo.validAuth().getLogin());
        authPage.enterPassword(DataHelper.AuthInfo.validAuth().getPass());
        authPage.signIn();
        mainScreenPage.checkMainScreenLoaded();
        mainScreenPage.isMainScreen();
    }

    @Test
    @Ignore
    @DisplayName("3.Вход в личный кабинет с пустым логином")
    @Description("Зависает. Не работает проверка всплывающих сообщений")
    public void testAuthEmptyLogin() {
        authPage.enterLogin(DataHelper.AuthInfo.emptyLogin().getLogin());
        authPage.enterPassword(DataHelper.AuthInfo.validAuth().getPass());
        authPage.signIn();
//   не работает матчер вслывающих сообщений никак
//        onView(withHint(R.string.empty_login_or_password)).inRoot(isPopupWindow()).
//                check(matches(isDisplayed()));
//        checkMessage(R.string.empty_login_or_password, true);
    }

    @Test
    @Ignore
    @DisplayName("4.Вход в личный кабинет с пустым паролем")
    @Description("Зависает. Не работает проверка всплывающих сообщений")
    public void testAuthEmptyPassword() {
        authPage.enterLogin(DataHelper.AuthInfo.validAuth().getLogin());
        authPage.enterPassword(DataHelper.AuthInfo.emptyPassword().getPass());
        authPage.signIn();
        checkMessage(R.string.empty_login_or_password, true);
    }

    @Test
    @Ignore
    @DisplayName("5.Вход в личный кабинет с неверным логином")
    @Description("Зависает. Не работает проверка всплывающих сообщений")
    public void testAuthInvalidLogin() {
        authPage.enterLogin(DataHelper.AuthInfo.wrongLogin().getLogin());
        authPage.enterPassword(DataHelper.AuthInfo.validAuth().getPass());
        authPage.signIn();
        checkMessage(R.string.wrong_login_or_password, true);
    }

    @Test
    @Ignore
    @DisplayName("6.Вход в личный кабинет с неверным паролем")
    @Description("Зависает. Не работает проверка всплывающих сообщений")
    public void testAuthInvalidPassword() {
        authPage.enterLogin(DataHelper.AuthInfo.validAuth().getLogin());
        authPage.enterPassword(DataHelper.AuthInfo.wrongPassword().getPass());
        authPage.signIn();
        checkMessage(R.string.wrong_login_or_password, true);
    }

    @Test
    @DisplayName("7.Выход из личного кабинета")
    public void testLogout() {
        authPage.enterLogin(DataHelper.AuthInfo.validAuth().getLogin());
        authPage.enterPassword(DataHelper.AuthInfo.validAuth().getPass());
        authPage.signIn();
        mainScreenPage.checkMainScreenLoaded();
        mainScreenPage.isMainScreen();
        mainScreenPage.logOut();
        authPage.isAuthScreen();
    }
}

File: app/src/androidTest/java/ru/iteco/fmhandroid/ui/AppActivityTest.kt
package ru.iteco.fmhandroid.ui


import android.view.View
import android.view.ViewGroup
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.*
import androidx.test.espresso.assertion.ViewAssertions.*
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import org.hamcrest.Description
import org.hamcrest.Matcher
import org.hamcrest.Matchers.allOf
import org.hamcrest.Matchers.`is`
import org.hamcrest.TypeSafeMatcher
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import ru.iteco.fmhandroid.R

@LargeTest
@RunWith(AndroidJUnit4::class)
class AppActivityTest {

    @Rule
    @JvmField
    var mActivityScenarioRule = ActivityScenarioRule(AppActivity::class.java)

    @Test
    fun appActivityTest() {
        val appCompatImageButton = onView(
            allOf(
                withId(R.id.main_menu_image_button), withContentDescription("Main menu"),
                childAtPosition(
                    allOf(
                        withId(R.id.container_custom_app_bar_include_on_fragment_main),
                        childAtPosition(
                            withClassName(`is`("android.widget.LinearLayout")),
                            0
                        )
                    ),
                    0
                ),
                isDisplayed()
            )
        )
        appCompatImageButton.perform(click())

        val materialTextView = onView(
            allOf(
                withId(android.R.id.title), withText("News"),
                childAtPosition(
                    childAtPosition(
                        withId(android.R.id.content),
                        0
                    ),
                    0
                ),
                isDisplayed()
            )
        )
        materialTextView.perform(click())

        val materialButton = onView(
            allOf(
                withId(R.id.edit_news_material_button),
                childAtPosition(
                    childAtPosition(
                        withId(R.id.container_list_news_include),
                        0
                    ),
                    3
                ),
                isDisplayed()
            )
        )
        materialButton.perform(click())

        val materialButton2 = onView(
            allOf(
                withId(R.id.add_news_image_view), withContentDescription("Add news button"),
                childAtPosition(
                    childAtPosition(
                        withClassName(`is`("android.widget.LinearLayout")),
                        1
                    ),
                    3
                ),
                isDisplayed()
            )
        )
        materialButton2.perform(click())
    }

    private fun childAtPosition(
        parentMatcher: Matcher<View>, position: Int
    ): Matcher<View> {

        return object : TypeSafeMatcher<View>() {
            override fun describeTo(description: Description) {
                description.appendText("Child at position $position in parent ")
                parentMatcher.describeTo(description)
            }

            public override fun matchesSafely(view: View): Boolean {
                val parent = view.parent
                return parent is ViewGroup && parentMatcher.matches(parent)
                        && view == parent.getChildAt(position)
            }
        }
    }
}
